### ISSUE TITLE
Generic classes, class methods, and `TypeVar`s with value restrictions don't seem to play well together

### ISSUE THREAD

**Issue by bwo** [python/mypy#16225]
**Bug Report**

Given the program below, mypy complains about `cls(s)` saying "main.py:12: error: Argument 1 to "Foo" has incompatible type "str"; expected "int"  [arg-type]" and about `cls(i)` saying "main.py:16: error: Argument 1 to "Foo" has incompatible type "int"; expected "str"  [arg-type]". Basically, whichever type is provided, mypy thinks the other type is expected.

This does not happen when simply using the constructor directly; it only seems to happen in class methods.

At the use site (eg if one does `reveal_type(Foo.build_str("hi"))`, mypy does not complain and in this case identifies the type as `Foo[str]`.

**To Reproduce**

```python
from typing import TypeVar, Optional, Generic, Self

T = TypeVar('T', str, int)


class Foo(Generic[T]):
    def __init__(self, t: T) -> None:
        self.t : T = t

    @classmethod
    def build_str(cls, s: str) -> Self:
        return cls(s)  # error

    @classmethod
    def build_int(cls, i: int) -> Self:
        return cls(i)  # error

Foo(1)  # no error
Foo("hi")  # no error
```

[playground link](https://mypy-play.net/?mypy=latest&python=3.11&gist=195ee38119f72af078bd987b2be0af44)

**Expected Behavior**

The program should be accepted.

**Actual Behavior**

```
foo.py:12: error: Argument 1 to "Foo" has incompatible type "str"; expected "int"  [arg-type]
foo.py:16: error: Argument 1 to "Foo" has incompatible type "int"; expected "str"  [arg-type]
Found 2 errors in 1 file (checked 1 source file)
```

**Your Environment**

- Mypy version used: 1.5.1
- Mypy command-line flags: none
- Python version used: 3.11

also, the same thing happens on the playground.

**Probably related**

The real reason I came across this is that I wanted to be able to do something like `-> Self[int]`; it's possible to do ` -> "Foo[int]"` for the return type, but that is obviously less precise. (In fact, `Self[int]` is not allowed.) In the process of fiddling with the above program I tried the below, thinking it might fix things:

```python
from typing import TypeVar, Optional, Generic, Self

T = TypeVar('T', str, int)


class Foo(Generic[T]):
    def __init__(self, t: T) -> None:
        self.t : T = t

    @classmethod
    def build_str(cls, s: str) -> "Foo[str]":
        return cls(s)  # error

    @classmethod
    def build_int(cls, i: int) -> "Foo[int]":
        return cls(i)  # error

Foo(1)  # no error
Foo("hi")  # no error
```

This actually produced *more* errors, of the same sort:

```
foo.py:12: error: Incompatible return value type (got "Foo[int]", expected "Foo[str]")  [return-value]
foo.py:12: error: Argument 1 to "Foo" has incompatible type "str"; expected "int"  [arg-type]
foo.py:16: error: Incompatible return value type (got "Foo[str]", expected "Foo[int]")  [return-value]
foo.py:16: error: Argument 1 to "Foo" has incompatible type "int"; expected "str"  [arg-type]
Found 4 errors in 1 file (checked 1 source file)

```

**Comment by erictraut** [comment]
Mypy is correct in emitting an error here. The type `Self` refers to a specialized type. For example, if you call `Foo[int].build_str("x")` and `build_str` is annotated to return `Self`, then the implementation of `build_str` must return a `Foo[int]` in that case.

Here's an alternative approach that does not violate any type rules:

```python
from typing import TypeVar, Generic

T = TypeVar("T", str, int)

S_Str = TypeVar("S_Str", bound="Foo[str]")
S_Int = TypeVar("S_Int", bound="Foo[int]")

class Foo(Generic[T]):
    def __init__(self, t: T) -> None:
        self.t: T = t

    @classmethod
    def build_str(cls: type[S_Str], s: str) -> S_Str:
        return cls(s)

    @classmethod
    def build_int(cls: type[S_Int], i: int) -> S_Int:
        return cls(i)

Foo(1)
Foo("hi")
```

**Comment by bwo** [comment]
using `Self` types seems to be an additional source of problems. This typechecks:

```python
from typing import TypeVar, Generic, Self                                                                                                                                                                                                                                                                                     
                                                                                                                                                                                                                                                                                                                              
T = TypeVar('T', str, int)                                                                                                                                                                                                                                                                                                    
                                                                                                                                                                                                                                                                                    
                                                                                                                                                                                                                                                                                                                              
class Foo(Generic[T]):                                                                                                                                                                                                                                                                                                        
    def __init__(self, t: T) -> None:                                                                                                                                                                                                                                                                                         
        self.t : T = t                                                                                                                                                                                                                                                                                                        
                                                                                                                                                                                                                                                                                                                              
    @classmethod                                                                                                                                                                                                                                                                                                              
    def build_str(cls: "type[Foo]", s: str) -> "Foo[str]":                                                                                                                                                                                                                                                                    
        return cls(s)                                                                                                                                                                                                                                                                                                         
                                                                                                                                                                                                                                                                                                                              
    @classmethod                                                                                                                                                                                                                                                                                                              
    def build_int(cls: "type[Foo]", i: int) -> "Foo[int]":                                                                                                                                                                                                                                                                    
        return cls(i)                                                                                                                                                                                                                                                                                                         
```

*However*, as one would expect, it is unsatisfactory in the presence of inheritance:


```python
class Bar(Foo[T]):                                                                                                                                                                                                                                                                                                            
    pass                                                                                                                                                                                                                                                                                                                      
                                                                                                                                                                                                                                                                                                                              
                                                                                                                                                                                                                                                                                                                              
reveal_type(Bar.build_str(""))                                                                                                                                                                                                                                                                                                
```                                                                                                                                                                                                                                                                                                                              
                                                                                                                                                                                                                                                                                                                              
this reveals a type of `Foo[str]`

**Comment by bwo** [comment]
the actual case is more complicated, and I'm not sure what I want can actually be made to work—the existence of separate methods which can alternately return one or another specialization of the generic type is kind of key. The issue clearly isn't what I thought it was, though.

**Comment by bwo** [comment]
fwiw, you can see here: https://github.com/python/mypy/issues/16225 a real-world (and public) example of the sort of thing I was coming up against.
